File: concrete/src/main/java/pcb/uwu/amount/finance/AUD.java
package pcb.uwu.amount.finance;

import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.unit.finance.CurrencyUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.utils.UnitAmountUtils.dividedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.minusAmount;
import static pcb.uwu.utils.UnitAmountUtils.multipliedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.plusAmount;

public class AUD extends Money {

	public static final CurrencyUnit CURRENCY = CurrencyUnit.of("AUD");

	// region constructor

	public AUD(Number value) {
		super(value, CURRENCY);
	}

	public AUD(String value) {
		super(value, CURRENCY);
	}

	public AUD(BigDecimal value) {
		super(value, CURRENCY);
	}

	public AUD(BigDecimalAmount amount) {
		super(amount, CURRENCY);
	}

	// endregion

	// region implement UnitAmount

	public AUD plus(AUD other) {
		return new AUD(getAmount().plus(other.getAmount().getValue()));
	}

	public AUD minus(AUD other) {
		return new AUD(getAmount().minus(other.getAmount().getValue()));
	}

	@Override
	public AUD plus(UnitAmount<CurrencyUnit> other, MathContext mathContext) {
		return new AUD(plusAmount(this, other, mathContext));
	}

	@Override
	public AUD minus(UnitAmount<CurrencyUnit> other, MathContext mathContext) {
		return new AUD(minusAmount(this, other, mathContext));
	}

	@Override
	public AUD multipliedBy(BigDecimal other, MathContext mathContext) {
		return new AUD(multipliedByScalar(this, other, mathContext));
	}

	@Override
	public AUD dividedBy(BigDecimal other, MathContext mathContext) {
		return new AUD(dividedByScalar(this, other, mathContext));
	}

	// endregion
}

File: concrete/src/main/java/pcb/uwu/amount/base/currency/AUD.kt
package pcb.uwu.amount.base.currency

import pcb.uwu.amount.base.Currency
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.CurrencyUnit
import pcb.uwu.unit.base.CurrencyUnit.Companion.AUD

class AUD : Currency
{
    @JvmOverloads
    constructor(amount: Number, magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, AUD)

    @JvmOverloads
    constructor(value: String, magnitude: Magnitude = NATURAL)
            : super(value, magnitude, AUD)

    // region UnitAmount

    override fun plus(currency: UnitAmount<CurrencyUnit>) =
        AUD(this.amount + (currency to AUD).amount)

    override fun minus(currency: UnitAmount<CurrencyUnit>) =
        AUD(this.amount - (currency to AUD).amount)

    override fun times(scalar: Number) =
        AUD(this.amount * scalar)

    override fun div(scalar: Number) =
        AUD(this.amount / scalar)

    // endregion
}